src/renderer/vt/paint.cpp

```diff
@@ -443,15 +443,23 @@ using namespace Microsoft::Console::Types;
     const bool useEraseChar = (optimalToUseECH) &&
                               (!_newBottomLine) &&
                               (!_clearedAllThisFrame);
+    const bool printingBottomLine = coord.Y == _lastViewport.BottomInclusive();
 
     // If we're not using erase char, but we did erase all at the start of the
     // frame, don't add spaces at the end.
     //
     // GH#5161: Only removeSpaces when we're in the _newBottomLine state and the
     // line we're trying to print right now _actually is the bottom line_
-    const bool removeSpaces = useEraseChar ||
-                              _clearedAllThisFrame ||
-                              (_newBottomLine && coord.Y == _lastViewport.BottomInclusive());
+    //
+    // GH#5291: DON'T remove spaces when the row wrapped. We might need those
+    // spaces to preserve the wrap state of this line, or the cursor position.
+    // For example, vim.exe uses "~    "... to clear the line, and then leaves

```

Comment:
We used to used to do something like "`80X 80C` (delete 80, move 80 right)" or "`K 80C`", right? Would that still work here? After all, we want the spaces to appear on the output but the wrap state to be maintained

is the real problem that we're not putting the cursor at the end of the line, and the spaces make up for that? and suppressing the spaces would mean we move the cursor wrong?

Reply:
(I'm worried about the accidental deoptimization in nor using ECH/EL, i guess.)

Reply:
if we just did the `^[[80X^[[80C`, then the line won't actually get wrapped. We need to emit an actual space there. It's a combo of both needing the spaces to trigger the wrapping logic, and needing the spaces to move the cursor.

Theoretically we could do `^[[79X^[[79C ` (with a space at the end), but that seems dirtier.

Reply:
Fair enough.

Reply:
I'm concerned about the efficiency of it, though. We realized a lot of gains recently simply by emitting fewer things between the PTY and the Terminal. It's a little dirtier, but it's 69 fewer bytes to send, parse, and react to. That doesn't sound like much for one line, but it could be easily magnified over multiple lines and redraws.

Reply:
So here's my thought - the client app was already emitting all these spaces intentionally. This was actually an optimization that conpty was doing _for them_, and an optimization that we were doing wrong (in this scenario). 

We'll still be sending the optimized sequences if the app does something like "Y &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; X", or if the app uses `^[[K` to clear the line (like a sane application should)

Reply:
Alright, then I'm fine with it. Thanks.